src/main/java/mil/dds/anet/threads/AccountDeactivationWorker.java

```diff
@@ -32,79 +33,94 @@
-    query.setEndOfTourDateEnd(warningDate);
-    List<Person> persons = AnetObjectEngine.getInstance().getPersonDao().search(query).getList();
-    Instant nextReminder =
-        daysTillNextWarning < 0 ? null : now.plus(daysTillNextWarning, ChronoUnit.DAYS);
-
-    // Send emails to let users know their account will soon be deactivated
-    // or deactivate accounts that reach the end-of-tour date
-    for (Person p : persons) {
-
-      // Skip inactive ANET users or users from ignored domains
-      if (p.getStatus() == PersonStatus.INACTIVE
-          || Utils.isDomainUserNameIgnored(p.getDomainUsername(), this.ignoredDomainNames)) {
-        continue;
+    final Instant now = Instant.now().atZone(DaoUtils.getDefaultZoneId()).toInstant();
+    final Instant latestWarningDate = now.plus(daysBeforeLatestWarning, ChronoUnit.DAYS);
+    query.setEndOfTourDateEnd(latestWarningDate);
+    final List<Person> persons =
+        AnetObjectEngine.getInstance().getPersonDao().search(query).getList();
+
+    // Send emails to let users know their account will soon be deactivated or deactivate accounts
+    // that reach the end-of-tour date
+    persons.forEach(p -> {
+      for (int i = 0; i < warningDays.size(); i++) {
+        final Integer warning = warningDays.get(i);
+        final Integer nextWarning = i == warningDays.size() - 1 ? null : warningDays.get(i + 1);
+        checkDeactivationStatus(p, warning, nextWarning, now);
       }
+    });
+  }
 
-      // Deactivate account as end-of-tour date has been reached
-      if (daysUntilEndOfTour == 0 && p.getEndOfTourDate().isBefore(now)) {
-        deactivateAccount(p);
-        continue;
-      }
+  private void checkDeactivationStatus(final Person person, final Integer daysBeforeWarning,
+      final Integer nextWarning, final Instant now) {
+    // Skip inactive ANET users or users from ignored domains
+    if (person.getStatus() == PersonStatus.INACTIVE
+        || Utils.isDomainUserNameIgnored(person.getDomainUsername(), this.ignoredDomainNames)) {
+      return;
+    }
 
-      // Send deactivation warning email
-      if (p.getEndOfTourDate().isBefore(warningDate) && p.getEndOfTourDate()
-          .isAfter(warningDate.minus(this.warningIntervalInMs, ChronoUnit.MILLIS))) {
-        sendDeactivationWarningEmail(p, nextReminder);
-      }
+    // Deactivate account as end-of-tour date has been reached
+    if (person.getEndOfTourDate().isBefore(now)) {
+      deactivateAccount(person);
+      return;
+    }
+
+    // Send deactivation warning email
+    final Instant warningDate = now.plus(daysBeforeWarning, ChronoUnit.DAYS);
+    final Instant nextReminder =
+        nextWarning == null ? null : now.plus(nextWarning, ChronoUnit.DAYS);
+    if (person.getEndOfTourDate().isBefore(warningDate) && person.getEndOfTourDate()
+        .isAfter(warningDate.minus(this.warningIntervalInSecs, ChronoUnit.SECONDS))) {
+      sendDeactivationWarningEmail(person, nextReminder);
     }
```

Comment:
Does this really prevent duplicate emails being sent, e.g. if ANET is restarted twice within a `checkIntervalInSecs` timeframe? And wouldn't this possibly skip sending emails if ANET is down for more than `checkIntervalInSecs`?

Reply:
Missed email warnings on server downtime do not seem like a big deal, but I am not the one to judge. Maybe @VassilIordanov can provide some input. I suggest pushing this and opening another issue to handle this later.

Reply:
After consulting @VassilIordanov the following was decided:
- A robust solution is desired. To tackle this an issue has been created (#2516)
- Push this PR after approval. It still is an improvement over the current solution and expands the unit tests.